### PR TITLE
Replace instance assignments with context managers in _grpc.impl

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Replace `DagsterInstance` assignments in grpc impl with use as a context manager. This improves performance over relying on garbage collection, per @gibsondan (see https://github.com/dagster-io/dagster/pull/13145/files#r1155100378)

## How I Tested These Changes

BK

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13337** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13337" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/13337?utm_source=stack-comment).

File: python_modules/dagster/dagster/_grpc/impl.py [hunk 1]
```diff
@@ -462,20 +462,22 @@ def get_partition_tags(
...
+            partition = partitions_def.get_partition(
+                partition_name, dynamic_partitions_store=instance
             )
-            return ExternalPartitionTagsData(name=partition.name, tags=tags)
+            with user_code_error_boundary(
+                PartitionExecutionError,
+                lambda: f"Error occurred during the evaluation of the `tags_for_partition` function for partitioned config on job '{job_def.name}'",
+            ):
+                tags = partitioned_config.get_tags_for_partition_key(
+                    partition.name, job_name=job_def.name
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_grpc/impl.py:478]: is the fix for the issue where instance needed to be passed in here going in separately?

File: python_modules/dagster/dagster/_grpc/impl.py [hunk 2]
```diff
@@ -527,47 +529,47 @@ def get_partition_set_execution_param_data(
...
+
+                with user_code_error_boundary(
+                    PartitionExecutionError, _error_message_fn(partition.name)
+                ):
+                    run_config = partitioned_config.get_run_config_for_partition_key(
+                        partition.name, instance
+                    )
+                    tags = partitioned_config.get_tags_for_partition_key(
+                        partition.name, instance, job_name=job_def.name
+                    )
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_grpc/impl.py:562]: likewise

**Review by gibsondan — APPROVED** [review]

File: python_modules/dagster/dagster/_grpc/impl.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_grpc/impl.py:562]: ? the instance is passed here

**Pull request merged**